### PR TITLE
ruler: improve error classification to support ruler failures

### PR DESCRIPTION
**Pull request by NickAnge**
#### What this PR does

Improve isOperatorControllable to:
- Support `vector contains metrics with the same labelset after applying alert labels` ([code](https://github.com/grafana/mimir-prometheus/blob/dccafb75552614127e769710d3a5545cd14bed66/rules/alerting.go#L444)) when alerts are generated with the same labels to avoid having duplicates alerts. This should count as `user`  controllable 
- Support grpc errors with no details. We count them as operator controllable

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`. If changelog entry is not needed, please add the `changelog-not-needed` label to the PR.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

---

> [!NOTE]
> Classifies gRPC errors without details as operator errors and rule evaluation failures (e.g., duplicate labelsets) as user errors, with updated tests and metrics checks for remote evaluation.
> 
> - **Ruler**
>   - **Error Classification**: 
>     - Treat gRPC errors without details as `operator` errors.
>     - Treat rule evaluation failures (e.g., duplicate labelsets after applying alert/rule labels) as `user` errors.
>   - **Metrics/Test Updates**:
>     - Integration test updates to assert `cortex_ruler_queries_failed_total` only when the query itself fails; no increment when query succeeds but evaluation fails.
>     - Add unit tests validating classification for undetailed gRPC errors and duplicate labelset errors; ensure detection via error string checks.
>   - **Changelog**: Document the new classifications for gRPC-without-details and duplicate-labelset evaluation failures.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 31a74d026de136c5f44f4f5b0c8ae07ac8edd509. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: pkg/ruler/compat.go [hunk 1]
```diff
@@ -498,12 +499,18 @@ func (c *rulerErrorClassifier) IsOperatorControllable(err error) bool {
 				return !mimirpb.IsClientErrorCause(errDetails.GetCause())
 			}
 		}
+		return true
```
[the remark this answers is not in the document]
Comment (NickAnge) [reply pkg/ruler/compat.go:513]: Fixed [faaee5a](https://github.com/grafana/mimir/pull/13586/commits/faaee5aecb15a0232b6780b75cbfa670904f0590)

**Review by tacole02 — COMMENTED** [review]
Changelog LGTM

File: pkg/ruler/compat_test.go [hunk 2]
```diff
@@ -1046,12 +1046,45 @@ func TestRulerErrorClassifier_IsOperatorControllable(t *testing.T) {
...
+		},
+		{
+			name:                   "gRPC Internal error (operator, local querier)",
+			err:                    status.New(codes.Internal, "codes.Internal").Err(),
+			expectedUserFailed:     false,
+			expectedOperatorFailed: true,
+			remoteQuerier:          false,
+		},
+
+		// Rule evaluation failures - duplicate labelsets
```
Comment (chencs) [inline pkg/ruler/compat_test.go:1066]: This change will also catch a similar error for recording rules ([code](https://github.com/grafana/mimir-prometheus/blob/6b991b2cf91e9283ed54b92bd6a61bf40dbde7fb/rules/recording.go#L107)). I don't think it's _necessary_ to add a test here for that error as well, but it might be nice for future understanding -- i.e., if someone comes looking to understand what errors this was anticipated to catch.

Comment (NickAnge) [reply pkg/ruler/compat_test.go:1066]: Yes. It should be for both recording and alert evaluation . Thats the reason I used `string.Contains` above.

I agree that this test is a nice to have to understand which errors the function will catch.

Comment (NickAnge) [reply pkg/ruler/compat_test.go:1066]: added [b870478](https://github.com/grafana/mimir/pull/13586/commits/b8704783a412e18d40d8e62591281fba24310b9f)

File: pkg/ruler/compat.go [hunk 3]
```diff
@@ -516,3 +523,7 @@ func (c *rulerErrorClassifier) IsOperatorControllable(err error) bool {
 	// Unknown errors
 	return true
 }
+
+func isUserRuleEvalFailure(err error) bool {
+	return strings.Contains(err.Error(), "vector contains metrics with the same labelset")
```
Comment (dimitarvdimitrov) [inline pkg/ruler/compat.go]: can you add a test to assert that that code still returns the same string when relabelling results in invalid alerts? I'm thinking some test either here or in the alertmanager package and then share a `const` with the error message in the helper function here and in the test.

a better alternative to that would be to expose that error as a sentinel from the alertmanager package and use `errors.Is` here. But i see how that's more work

Comment (NickAnge) [reply pkg/ruler/compat.go]: > a better alternative to that would be to expose that error as a sentinel from the alertmanager package and use errors.Is here. But i see how that's more work

I thought about and I will do it a separate PR . When I raised it I wanted to merge it last week, so it was part of the latest release. Would it be ok for you if I handle this in a separate PR ?

> can you add a test to assert that that code still returns the same string when relabelling results in invalid alerts? I'm thinking some test either here or in the alertmanager package and then share a const with the error message in the helper function here and in the test.

Do you mean to check, if the `isUserRuleEvalFailure` returns the same error string for errors that do not match 
```
strings.Contains(err.Error(), "vector contains metrics with the same labelset")
```
?

Comment (dimitarvdimitrov) [reply pkg/ruler/compat.go]: no i mean add a test in mimir that asserts that the code here https://github.com/grafana/mimir-prometheus/blob/dccafb75552614127e769710d3a5545cd14bed66/rules/alerting.go#L444 returns the error string that mimir expects on that kind of failures. 

right now if mimir-prometheus changes its error string, this code might break silently (not necessarily silently you added some tests, but it might be a while until someone catches the bug while debugging)

File: pkg/ruler/compat.go [hunk 4]
```diff
@@ -43,6 +44,12 @@ const (
 	failureReasonClientError = "client_error"
 )
 
+// TODO: remove this once these errors are populated as sentinenls at
```
Comment (dimitarvdimitrov) [inline pkg/ruler/compat.go]:
```suggestion
// TODO: remove this once these errors are populated as sentinenls at mimir-prometheus
```

?

Comment (NickAnge) [reply pkg/ruler/compat.go]: [31a74d0](https://github.com/grafana/mimir/pull/13586/commits/31a74d026de136c5f44f4f5b0c8ae07ac8edd509)

File: pkg/ruler/compat.go (hunk 3 above)
Comment (NickAnge) [reply pkg/ruler/compat.go]: Done [31a74d0](https://github.com/grafana/mimir/pull/13586/commits/31a74d026de136c5f44f4f5b0c8ae07ac8edd509)
